clang/lib/AST/ExprConstant.cpp

```diff
@@ -14638,6 +14649,8 @@ bool FloatExprEvaluator::VisitCallExpr(const CallExpr *E) {
     return true;
   }
 
+  case Builtin::BIfmin:
+  case Builtin::BIfminf:
```

Comment:
Don't make these evaluate in `constexpr`. Not only does this extend C (not a stated intent of the patch), it will cause accidental dependencies that break when `-fno-builtin` is used.

Reply:
These have to be made `constexpr`, at least in C++23. C++ stdlibs can't override them to make them call the `__builtin_` versions, so Clang has to handle that.

Reply:
Makes me wonder how far C++ stdlibs will get if they included the C library headers in an inline namespace under the global namespace (and count on `extern "C"` to handle the linkage names).

Reply:
How would that help?

Reply:
> How would that help?

I guess it doesn't... The functions would be ambiguous when called...

Reply:
I think we'll be missing at least "pedantic" diagnostics if we don't restrict the `constexpr`-ness to C++23 and up. So we do need that restriction.

The interaction with `-fno-builtin[-*]` may need more discussion. At least the documentation for that option would need to be updated if the C++23 `constexpr` math would break with it.

Reply:
At the Clang C/C++ Language Workgroup call on 2024-05-15, it was agreed that `-fno-builtin[-*]` should remain capable of disabling built-in treatment even when that treatment would be required for conformance w.r.t. `constexpr` behaviour of functions shared with C.

For me, the first expectation for this PR (on that subject) is an update to the documentation of `-fno-builtin[-*]`.

@AaronBallman @ldionne, I am no sure how we want to handle this w.r.t. documenting the C++23 implementation status. Does it go in both the Clang documentation (for the functions shared with C) and the libc++ documentation (for the overloads added by C++)?

Reply:
I don't think anything should happen to the libc++ documentation in this patch. We can update the implementation to behave the same across the different overloads in a follow-up and update our documentation then.